Add security token auth and credential loading from msticpyconfig.yaml to SplunkUploader

Tatsuya-hasegawa opened the pull request:
Hi, 

I have added the auth features to SplunkUploader as same as Splunk QueryProvider.
- security token authentication
- credential loading from msticpyconfig.yaml
It was just to wrap "splunk_driver.py" 's connect() method.
I have tested in my Splunk Enterprise instance whether this code is fine.

Related issue.
https://github.com/microsoft/msticpy/issues/693

This PR resolves the security concern about the current hard coding style of username and password.

Thank you !
Best regards,

ianhelle commented on docs/source/data_acquisition/UploadData.rst:
```diff
@@ -81,6 +81,7 @@ Instantiating the Splunk uploader
 The first step in uploading data is to instantiate an uploader for the location we wish to upload data to.
 For Splunk there are three parameters that need to be passed at this stage, the Splunk host name, a username,
 and a password. You can also pass a parameter for ``port``, by default this value is 8089.
+In addition, The security auth token of ``splunkToken`` can be also passed instead of username and password as same as Splunk QueryProvider.
```
Can you change to `splunk_token` (to follow python param naming conventions.
Actually, can it just be `token` ? Since we already know that it's for Splunk. I think `token` is what we used for the data provider.

Tatsuya-hasegawa replied:
Yes, Of course I can. I did it in the new commit.  The name token has already used as session token in the driver's connect parameter, so I used bearer_token instead of splunkToken referring to Splunk python SDK as following.
https://github.com/splunk/splunk-sdk-python/blob/master/splunklib/binding.py#L523

Thank you for teaching me !

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -17,19 +17,20 @@
+        self.host = kwargs.get("host")
+        self.password = kwargs.get("password")
+        self.username = kwargs.get("username")
+        self.splunkToken = kwargs.get("splunkToken")
```
Same here - rename param and attribute to use snake casing. 
param can be `token` but attribute could be splunk_token or token

Tatsuya-hasegawa replied:
Checked and Done!

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -40,12 +41,28 @@ def __init__(self, username: str, host: str, password: str, **kwargs):
+                    port=self.port,
+                )
+            else:
+                print("username and splunkToken are missing.")
```
I'm not a big fan of printing out stuff without any user control.
It might better to add a logger to the module and output the message as an info message.
```python3
import logging
...
logger = logging.getLogger(__name__)

# then in the code 
logger.info("username/token not given as parameters - using config settings.")
```

Tatsuya-hasegawa replied:
Checked and Done!

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -40,12 +41,28 @@ def __init__(self, username: str, host: str, password: str, **kwargs):
+                print("Credential loading from config file.")
+                self.driver.connect()
+        else:
+            print("Credential loading from config file.")
```
and here

Tatsuya-hasegawa replied:
Checked and Done!

Tatsuya-hasegawa commented:
@ianhelle

I have updated all the points! 
I appreciate your comment. 

Unfortunately, Splunk python SDK don't use snake case but camel case, it's unlucky for me.
I added a code snippet to splunk_driver.py in order to  rollback bearer_token to splunkToken in front of splunk client's connect(). 

These codes's splunk connection tests were ok for both query provider and uploader.

>Yes, Of course I can. I did it in the new commit.  The name "token" has already used as session token in the driver's connect parameter, so I used bearer_token instead of splunkToken referring to Splunk python SDK as following.
> https://github.com/splunk/splunk-sdk-python/blob/master/splunklib/binding.py#L523

Tatsuya-hasegawa commented:
Hi @ianhelle 

I recognize that the requested modification has been completed. :)
Would you teach me the rest task for me, if that is not enough ?
I thought processing workflows need your approval.

Best regards,

ianhelle commented:
> Hi @ianhelle
> 
> I recognize that the requested modification has been completed. :) Would you teach me the rest task for me, if that is not enough ? I thought processing workflows need your approval.
> 
> Best regards,

Hey sorry. I've been pulled away on other work. The changes look good. Just running a build and will complete if everything looks OK

ianhelle commented:
Seems there are a few failing tests and linter failures. See if you can fix these (they look simple) and ping me if not.
It would be worth installing our pre-commit script - it will run many of the checks before letting you commit.
```bash
pip install pre-commit
# from the repo root
pre-commit install
```

Tatsuya-hasegawa commented:
@ianhelle 

I have modified my lint error.
And this is my pre-check result. It looks passed.
In my laptop, it was needed to change "isort" rev version to 5.12.0 in .pre-commit.yaml.
I'm sorry for ignoring to treat this error until now.
I believe I can do pre-check correctly from now on.
 
```
(base) hacket@hackeTlab msticpy % git commit -m  "modify just lints"
[INFO] Initializing environment for https://github.com/pycqa/isort.
[INFO] Installing environment for https://github.com/pycqa/isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pycqa/pydocstyle.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for local.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Check Yaml...............................................................Passed
Check JSON...........................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
check_reqs_all...........................................................Passed
[add_token_auth_to_splunk_uploader 4fbaa95f] modify just lints
 2 files changed, 4 insertions(+), 4 deletions(-)
```

Thank you for teaching.  😄

ianhelle commented:
Still some tests failing. I think instead of the check for arg_dict["bearer_token"] you should be prob be using
arg_dict.get("bearer_token")
There are also some very simple linter fixes - running black on the uploader file should fix this. The flake8 check (about additional blank lines will probably be fixed by this).
Apologies for delay in getting these run.

Tatsuya-hasegawa commented:
@Ianhelle 

Thank you. I have modified these points!
Best regards,

`(base) hacket@hackeTlab msticpy % git commit -m "a few fix for pytest result." 
Check Yaml...........................................(no files to check)Skipped
Check JSON...........................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
check_reqs_all...........................................................Passed
[add_token_auth_to_splunk_uploader 896da43a] a few fix for pytest result.
 2 files changed, 2 insertions(+), 1 deletion(-)`

ianhelle commented:
@Tatsuya-hasegawa - thanks for sticking with this. Apologies for taking so long to get it completed

Tatsuya-hasegawa commented:
Thank you for merging this !
